### PR TITLE
fix(jqLite): make removeData() not remove event handlers

### LINKED ISSUE #15869: jqLite#removeData shouldn't touch the event handlers' storage
**I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] other (Please do not submit support requests here (see above))

**Current behavior:**
`jqLite#removeData` currently:
1. Clears the element data.
2. Detaches event handlers.

**Expected / new behavior:**
The second behavior is incorrect; the only data touched should be user-provided data, i.e. those set via `elem.data(key, value)`. That's how jQuery behaves.

If complete deallocation of event handlers is required as well, one can use:
```js
jqLite.cleanData(elements);
```

Note that currently `jqLite.cleanData` defers to `jqLiteRemoveData` under the hood so its implementation will have to be changed when `jqLiteRemoveData` is fixed.

**Minimal reproduction of the problem with instructions:**
AngularJS + jqLite (handler removed): http://plnkr.co/edit/XBw30umyz68UnERE9LyS?p=preview
AngularJS + jQuery (handler not removed): http://plnkr.co/edit/sdo9ShVA66yuEIIEvlII?p=preview

**Angular version:** 1.x.y

1.6.3, snapshot as well

**Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

All

**Anything else:**

This will require a breaking change in jqLite. Setting to 1.7.0.

### PR DESCRIPTION
**Pull request by mgol**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Bug fix but a breaking change as well.

**What is the current behavior? (You can also link to an open issue here)**

`removeData()` invoked on an element removes its event handlers as well.

**What is the new behavior (if this is a feature change)?**

`removeData()` removes just user-set `data`, not internal data storage used for event handling. This aligns jqLite with jQuery.

**Does this PR introduce a breaking change?**

Yes.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- ~~Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated~~ (the current behavior is not documented)
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

Fixes #15869

### CODE REVIEW

File: test/jqLiteSpec.js [hunk 1]
```diff
@@ -420,6 +420,29 @@ describe('jqLite', function() {
...
+        log += 'click;';
+      });
+
+      elm.removeData();
+      browserTrigger(a, 'click');
+      expect(log).toBe('click;');
+    });
+
+
+    it('should allow to set data after removeData() with event handlers present', function() {
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: What about "with event handlers not present"?

**Review by gkalpak — CHANGES_REQUESTED** [review]
No specs for `cleanData()` 😱
Also, in the commit message, I think you should change `angular.element.cleanData([elem])` to `angular.element.cleanData(elem)`.

**Comment by mgol** [comment]
> No specs for cleanData() 😱

Good point. Although this method is still not officially documented in jQuery although it seems we'd like to change it.

> Also, in the commit message, I think you should change angular.element.cleanData([elem]) to angular.element.cleanData(elem).

The confusion comes from the fact jqLite code refes to bare DOM nodes as `element`. `cleanData` accepts an array-like of bare DOM nodes, not just jqLite objects so in this context `elem` is the correct choice as you noticed.

File: src/jqLite.js [hunk 2]
```diff
@@ -615,6 +633,7 @@ forEach({
   cleanData: function jqLiteCleanData(nodes) {
     for (var i = 0, ii = nodes.length; i < ii; i++) {
       jqLiteRemoveData(nodes[i]);
+      jqLiteOff(nodes[i]);
```
[the remark this answers is not in the document]
Comment (mgol) [reply src/jqLite.js]: This should be handled in the `cleanData` patch that used to be applied to jQuery only and that I moved to apply to jqLite as well.

File: src/jqLite.js [hunk 3]
```diff
@@ -310,6 +310,28 @@ function jqLiteDealoc(element, onlyDescendants) {
   }
 }
 
+function isEmptyObject(obj) {
+  var name;
+
+  for (name in obj) {
```
[the remark this answers is not in the document]
Comment (mgol) [reply src/jqLite.js]: That's taken from jQuery data handling:
https://github.com/jquery/jquery/blob/3.3.1/src/event.js#L290-L293
https://github.com/jquery/jquery/blob/3.3.1/src/data/Data.js#L142-L153

**Comment by gkalpak** [comment]
BTW, should [this][1] also be changd to `cleanData()`?

[1]: https://github.com/mgol/angular.js/blob/acbeb82b24ec022b713aaf91ea45f9dd6b42b139/test/helpers/testabilityPatch.js#L23

**Comment by mgol** [comment]
> BTW, should [this](https://github.com/mgol/angular.js/blob/acbeb82b24ec022b713aaf91ea45f9dd6b42b139/test/helpers/testabilityPatch.js#L23) also be changd to `cleanData()`?

Perhaps. `.empty()` will already call it for all descendants and `.removeData()` will remove user data on `body` so the only thing left are event handlers on `body`. But we're already not clearing handlers on `window` & `document` so it's not a full cleanup.

File: test/jqLiteSpec.js [hunk 4]
```diff
@@ -420,6 +420,87 @@ describe('jqLite', function() {
...
+      var selected = jqLite([a, b, c]);
+
+      var log = '';
+      var elm = jqLite(b);
+      elm.on('click', function() {
+        log += 'click;';
+      });
+      jqLite.cleanData(selected);
+
+      browserTrigger(a, 'click');
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: Did you mean `b`? :grin:

File: test/jqLiteSpec.js [hunk 5]
```diff
@@ -420,6 +420,87 @@ describe('jqLite', function() {
...
+      elm.on('click', function() {
+        log += 'click;';
+      });
+
+      selected.data('prop', 'value');
+      jqLite(a).data('prop', 'new value');
+
+      jqLite.cleanData(selected);
+
+      browserTrigger(a, 'click');
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: Did you mean `b`? :grin:

File: test/jqLiteSpec.js [hunk 6]
```diff
@@ -420,6 +420,87 @@ describe('jqLite', function() {
...
+      expect(log).toBe('');
+
+      expect(jqLite(a).data('prop')).toBeUndefined();
+      expect(jqLite(b).data('prop')).toBeUndefined();
+      expect(jqLite(c).data('prop')).toBeUndefined();
+    });
+
+    it('should remove user data & event handlers on cleanData()', function() {
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: This description belongs to the test above and vice versa.

**Review by gkalpak — APPROVED** [review]

File: test/jqLiteSpec.js (hunk 4 above)
Comment (mgol) [reply test/jqLiteSpec.js]: Good eye. :)

**Comment by mgol** [comment]
PR updated once again.

**Review by jbedard — APPROVED** [review]

**Pull request merged**